### PR TITLE
Add a Redis cache

### PR DESCRIPTION
**Pull request by maximemenager**

### CODE REVIEW

**Comment by sblaisot** [comment]
Please also update example config file https://github.com/cozy/cozy-apps-registry/blob/master/cozy-registry.example.yml

File: main.go [hunk 1]
```diff
@@ -182,6 +185,32 @@ func useConfig(cmd *cobra.Command) (err error) {
 		return fmt.Errorf("Failed to read cozy-apps-registry configuration %q: %s",
 			cfgFile, err)
 	}
+
+	// Create cache
+	if viper.GetString("redis.url") != "" {
+		redisCacheVersionsLatest := redis.NewClient(&redis.Options{
+			Addr:     viper.GetString("redis.url"),
+			Password: viper.GetString("redis.password"),
+			DB:       0,
```
Comment (sblaisot) [inline main.go]: Please make db number configurable

File: main.go [hunk 2]
```diff
@@ -182,6 +185,32 @@ func useConfig(cmd *cobra.Command) (err error) {
...
+		})
+		redisCacheVersionsList := redis.NewClient(&redis.Options{
+			Addr:     viper.GetString("redis.url"),
+			Password: viper.GetString("redis.password"),
+			DB:       1,
```
Comment (sblaisot) [inline main.go]: Same here.

File: cache/lru.go [hunk 3]
```diff
@@ -88,22 +99,24 @@ func (c *Cache) Get(key Key) (value Value, ok bool) {
...
+func (c *LRUCache) Remove(key Key) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 	if ele, hit := c.cache[key]; hit {
 		c.removeElement(ele)
 	}
 }
 
-func (c *Cache) removeOldest() {
+func (c *LRUCache) removeOldest() {
```
Comment (sblaisot) [inline cache/lru.go]: I think it's better to keep coherent functions between LRUCache and RedisCache so either remove this function or implement it in RedisCache

File: cache/lru.go [hunk 4]
```diff
@@ -88,22 +99,24 @@ func (c *Cache) Get(key Key) (value Value, ok bool) {
...
 	if ele := c.ll.Back(); ele != nil {
 		c.removeElement(ele)
 	}
 }
 
-func (c *Cache) removeElement(e *list.Element) {
+func (c *LRUCache) removeElement(e *list.Element) {
```
Comment (sblaisot) [inline cache/lru.go]: same here.

File: cache/lru.go (hunk 3 above)
Comment (nono) [reply cache/lru.go]: In go, the methods that start with an upper-case letter are exported (they can be used from other modules), and methods that start with a lower-case letter are private. Here, `removeOldest` and `removeElement` are just two helpers. The two cache implementations (Redis and LRU) expose the same interface.

Comment (sblaisot) [reply cache/lru.go]: Oh, ok, thanks.

File: main.go [hunk 5]
```diff
@@ -187,25 +187,44 @@ func useConfig(cmd *cobra.Command) (err error) {
...
+			PoolSize:           viper.GetInt("redis.pool_size"),
+			PoolTimeout:        viper.GetDuration("redis.pool_timeout"),
+			IdleTimeout:        viper.GetDuration("redis.idle_timeout"),
+			IdleCheckFrequency: viper.GetDuration("redis.idle_check_frequency"),
+		}
+		optsList := redisOpts
+		optsLatest := redisOpts
+
+		optsList.DB = viper.GetInt("redis.databases.versionsList")
+		optsLatest.DB = viper.GetInt("redis.databases.versionsLatest")
```
Comment (nono) [inline main.go]: You are sure that the `optsList.DB` will be taken in account? It looks like `optsList` and `optsLatest` are pointers to the same struct, and changes for one impacts the other.

Comment (maximemenager) [reply main.go]: Nice pick, I read the doc too fast. Struct copy like this works only with primitive types

**Pull request merged**